### PR TITLE
4039 Fix inconsistent results when ordering by date sorting keys.

### PR DESCRIPTION
**Pull request by albertisfu**
This pull request addresses issue #4032, which caused inconsistent sort orders when using `date_filed` for ascending sorts.

- The problem was directly related to queries that matched a docket by a RECAPDocument field in the `d` and `r` search types.

The original query structure was:

 ```
bool
  should
    function_score to nullify the child score
      has_child_query
   function_score to compute the score based on the docket date_filed
      docket_query
```

After profiling the queries, I realized that when a docket is matched by one of its child documents, the docket doesn't match the query. The docket function score doesn't contribute to the final document score.
The solution was to apply the function score to the whole query, so it applies regardless of the query that matched the document.

```
function_score to compute the score based on the docket `date_filed`
  bool
    should
      has_child_query
      docket_query
```

- The second issue solved in this PR was related to using the same function score to sort documents by `date_filed` or `entry_date_filed` in descending order, which had no issue since the document date is used directly to compute the score. 
 However, when sorting in ascending order, the score is computed as `current_time - date_filed_time`, causing older documents to have a greater score. This could lead to pagination inconsistencies over time as `current_time` was based on milliseconds. As time passed between requests, scores could change between pages. 
 In the front end, this is not an issue since all documents change their score equally. However, for cursor pagination, this could lead to inconsistencies across pages as these scores are used as a `search_after` parameter, potentially resulting in missing documents across pages.

  - To solve the issue, a `request_date` is retrieved on the first-page request, always set to the current day at midnight, ensuring all requests in a day have the same `request_date` and scores remain stable throughout the day due to `date_filed` and `entry_date_filed` are also day granular.
  - The second tweak was to pass this `request_date` to the cursor and use it to compute scores across pagination using the same cursor. This resolves the problem of score stability if a request is made on two different days, which might seem unusual but can be common for requests made close to day transitions.

Let me know what you think.

### CODE REVIEW

File: cl/lib/elasticsearch_utils.py [hunk 1]
```diff
@@ -751,16 +754,34 @@ def build_custom_function_score_for_date(
...
+    midnight_current_date = (
+        datetime.datetime.combine(default_current_date, datetime.time())
+        if default_current_date
+        else None
+    )
+    default_current_time = (
+        int(midnight_current_date.timestamp() * 1000)
+        if midnight_current_date
+        else None
+    )
```
Comment (ERosendo) [inline cl/lib/elasticsearch_utils.py:771]: I think we can simplify this logic by using a single if statement

```suggestion
    midnight_current_date = None
    default_current_time = None
    if default_current_date:
        midnight_current_date = datetime.datetime.combine(
            default_current_date, datetime.time()
        )
        default_current_time = int(midnight_current_date.timestamp() * 1000)
```

File: cl/api/pagination.py [hunk 2]
```diff
@@ -92,17 +93,29 @@ class ESCursorPagination(BasePagination):
...
+        self.cursor = self.decode_cursor(request)
+
+        # Set the request date from the cursor or provide an initial one if
+        # this is the first page request.
+        self.request_date = (
+            self.cursor.request_date
+            if self.cursor
+            else datetime.datetime.now().date()
+        )
+        return self.request_date
```
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/api/pagination.py:111]: Yeah, I agree with the name change!

File: cl/search/tests/tests_es_recap.py [hunk 3]
```diff
@@ -3090,10 +3176,153 @@ def test_date_filed_sorting_function_score(self) -> None:
...
+            ]:
+                test_param_copy = test_param.copy()
+                test_param_copy["type"] = search_type
+                all_tests.append(test_param_copy)
+
+        for search_params in all_tests:
+            with self.subTest(
+                search_params=search_params, msg="Test stable scores."
+            ):
+                original_datetime = now().replace(day=1, hour=5, minute=0)
```
Comment (ERosendo) [inline cl/search/tests/tests_es_recap.py:3249]: I think we can get this variable out of the loop, right? We don't need to recompute it for each subtest.

File: cl/search/tests/tests_es_recap.py [hunk 4]
```diff
@@ -3090,10 +3176,153 @@ def test_date_filed_sorting_function_score(self) -> None:
...
+                # Two first-page requests (no cursor) made on the same day will now
+                # have consistent scores because scores are now computed based on the
+                # same day's date.
+                with time_machine.travel(original_datetime, tick=False):
+                    r = self.client.get(
```
Comment (ERosendo) [inline cl/search/tests/tests_es_recap.py:3254]: We can potentially improve code reusability by implementing the [shift](https://github.com/adamchainz/time-machine?tab=readme-ov-file#shiftdelta) helper function. This function encapsulates the logic to move the current time by a given offset. By using this function before triggering an API request, we can avoid repeating the "time_machine" contexts throughout the code.

We would need to apply the following refactor to the code:

```python
original_datetime = now().replace(day=1, hour=5, minute=0)

for search_params in all_tests:
   with self.subTest(
       search_params=search_params, msg="Test stable scores."
   ), time_machine.travel(original_datetime, tick=False) as traveler:
   
        # first API request 
        date_score_1 = r.data["results"][0]["date_score"]


       traveler.shift(datetime.timedelta(minutes=1))
       # 2nd API request
       date_score_2 = r.data["results"][0]["date_score"]

       ...
```  

what do you think?

**Review by ERosendo — APPROVED** [review]
I've tested the code using several cloned dockets. The results within the API response and the corresponding frontend results consistently match. We can merge the PR after addressing the comments

File: cl/lib/elasticsearch_utils.py (hunk 1 above)
Comment (albertisfu) [reply cl/lib/elasticsearch_utils.py]: changed, thanks for the suggestion!

File: cl/api/pagination.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/api/pagination.py]: Done!

File: cl/search/tests/tests_es_recap.py (hunk 3 above)
Comment (albertisfu) [reply cl/search/tests/tests_es_recap.py]: Yeah, thanks for noticing it. I have moved the variable out of the loop.

File: cl/search/tests/tests_es_recap.py (hunk 4 above)
Comment (albertisfu) [reply cl/search/tests/tests_es_recap.py]: Yeah I agree, this is a perfect use case for `shift`.

**Comment by albertisfu** [comment]
Thanks, @ERosendo I've applied your suggestions!

**Pull request merged**